include/grpc++/impl/codegen/async_stream.h

```diff
@@ -505,6 +640,40 @@ class ServerAsyncReaderWriter final
     call_.PerformOps(&write_ops_);
   }
 
+  void Write(const W& msg, WriteOptions options, void* tag) override {
```

Comment:
Just be wary here: too much abstraction is going to hurt readability and performance.

I can definitely see adding an 'EnsureInitialMetadataSent()' method that takes over this stanza however:
```
   if (!ctx_->sent_initial_metadata_) {
      write_ops_.SendInitialMetadata(ctx_->initial_metadata_,
                                     ctx_->initial_metadata_flags());
      if (ctx_->compression_level_set()) {
        write_ops_.set_compression_level(ctx_->compression_level());
      }
      ctx_->sent_initial_metadata_ = true;
    }
```